stl/inc/xstring

```diff
@@ -2815,6 +2815,51 @@ private:
     }
 
 public:
+#ifdef __cpp_lib_containers_ranges
+    template <_Container_compatible_range<_Elem> _Rng>
+    constexpr basic_string(from_range_t, _Rng&& _Range, const _Alloc& _Al = _Alloc())
+        : _Mypair(_One_then_variadic_args_t{}, _Al) {
+        auto&& _Alproxy = _GET_PROXY_ALLOCATOR(_Alty, _Getal());
+        _Container_proxy_ptr<_Alty> _Proxy(_Alproxy, _Mypair._Myval2);
+        _Tidy_init();
+
+        if constexpr (_RANGES sized_range<_Rng>) {
+            const auto _Count = _Convert_size<size_type>(_To_unsigned_like(_RANGES size(_Range)));
+
+            if constexpr (contiguous_iterator<decltype(_RANGES _Ubegin(_Range))> && //
+                          same_as<_RANGES range_value_t<_Rng>, _Elem>) {
+                if (_Count != 0) {
+                    const auto _Ptr = _STD to_address(_RANGES _Ubegin(_Range));
+                    assign(_Ptr, _Count);
+                }
+            } else {
+                reserve(_Count);
+                _Tidy_deallocate_guard<basic_string> _Guard{this};
+                auto _First      = _RANGES _Ubegin(_Range);
+                const auto _Last = _RANGES _Uend(_Range);
+
+                for (; _First != _Last; ++_First) {
+                    _Push_back_unchecked(*_First);

```

Comment:
I don't love that this seems like it's doing a lot of unnecessary work when `_Traits::assign` is noexcept (a common enough case).

It would result in N less calls to `_Trait::assign`, and N less increments.

```cxx
if constexpr (noexcept(_Traits::assign(_Ptr[0], *_First))) {
    reserve(_Count);
    _My_data._Mysize = _Count;
    _Traits::assign(_Ptr[_My_data._Mysize], _Elem());
    for (auto _Dest = _Ptr; _Count != 0; --_Count, (void)++_First, ++_Dest) {
        _Traits::assign(_Dest, *_First);
    }
}
```

Reply:
Burninating per previous comment. (Aside: `char_traits` members may never emit exceptions, which fact we recently clarified in LWG-3518.)